src/tactic/apply_fun.lean

```diff
@@ -45,19 +45,75 @@ do {
   try $ tactic.dsimp_hyp hyp simp_lemmas.mk [] { eta := false, beta := true }
 }
 
+/--
+Attempt to "apply" a function `f` represented by the argument `e : pexpr` to the goal.
+
+If the goal is of the form `a ≠ b`, we obtain the new goal `f a ≠ f b`.
+If the goal is of the form `a = b`, we obtain a new goal `f a = f b`, and a subsidiary goal
+`injective f`.
+(We attempt to discharge this subsidiary goal automatically, or using the optional argument.)
+If the goal is of the form `a ≤ b` (or similarly for `a < b`), and `f` is an `order_iso`,
+we obtain a new goal `f a ≤ f b`.
+-/
+meta def apply_fun_to_goal (e : pexpr) (lem : option pexpr) : tactic unit :=
+do t ← target,
+  match t with
+  | `(%%l ≠ %%r) := (do
+      e' ← to_expr ``(ne_of_apply_ne %%e),
+      apply e',
+      skip)
+  | `(¬%%l = %%r) := (do
+      e' ← to_expr ``(ne_of_apply_ne %%e),
+      apply e',
+      skip)
+  | `(%%l = %%r) := (do
+      to_expr ``(%%e %%l), -- build and discard an application, to fill in implicit arguments
+      n ← get_unused_name `inj,
+      to_expr ``(function.injective %%e) >>= assert n,
+      -- Attempt to discharge the `injective f` goal
+      assumption <|>
+        (to_expr ``(equiv.injective) >>= apply >> skip) <|>
+        (lem.mmap (λ l, to_expr l >>= apply) >> skip),
+      -- Return to the main goal
+      swap,
```

Comment:
I'm confused. What goals are expected here? Are you expecting to see the equality and possibly the injectivity goal, if the attempt to discharge made progress but didn't finish?

If this is right: you need to focus at the beginning of the tactic, since if you have multiple goals and clear the injectivity goal this `swap` will give you something unexpected.

But also the discharging attempt is odd, it will fail if it makes no progress but doesn't mind partial progress from applying `lem`. I'd kind of expect it to never fail (leaving an injectivity goal) or to fail whenever it doesn't fully solve the side condition.

Reply:
Ah, I hadn't considered the possiblity that `lem` would make partial progress.

Reply:
@robertylewis, hopefully this is now better and better commented.